modules/sitemaps/sitemap-librarian.php

```diff
@@ -0,0 +1,593 @@
+	 */
+	public function query_latest_approved_comment_time_on_post( $post_id ) {
+		global $wpdb;
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT MAX(comment_date_gmt)
+					FROM $wpdb->comments
+					WHERE comment_post_ID = %d AND comment_approved = '1' AND comment_type=''",
+				$post_id
+			)
+		);
+	}
+
+	/**
+	 * Retrieve an array of image posts sorted by ID.
+	 *
+	 * More precisely, returns the smallest $num_posts image posts
+	 * (measured by ID) which are larger than $from_id.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $from_id Greatest lower bound of retrieved image post IDs.
+	 * @param int $num_posts Largest number of image posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_images_after_id( $from_id, $num_posts ) {
+		global $wpdb;
+
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+					FROM $wpdb->posts
+					WHERE post_type='attachment'
+						AND post_mime_type IN ('image/jpeg','image/png','image/gif')
+						AND ID>%d
+					ORDER BY ID ASC
+					LIMIT %d;",
+				$from_id,
+				$num_posts
+			)
+		); // WPCS: db call ok; no-cache ok.
+	}
+
+	/**
+	 * Retrieve an array of video posts sorted by ID.
+	 *
+	 * More precisely, returns the smallest $num_posts video posts
+	 * (measured by ID) which are larger than $from_id.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $from_id Greatest lower bound of retrieved video post IDs.
+	 * @param int $num_posts Largest number of video posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_videos_after_id( $from_id, $num_posts ) {
+		global $wpdb;
+
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+					FROM $wpdb->posts
+					WHERE post_type='attachment'
+						AND post_mime_type IN ('video/mpeg','video/wmv','video/mov')
+						AND ID>%d
+					ORDER BY ID ASC
+					LIMIT %d;",
+				$from_id,
+				$num_posts
+			)
+		); // WPCS: db call ok; no-cache ok.
+	}
+
+	/**
+	 * Retrieve an array of published posts from the last 2 days.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $num_posts Largest number of posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_most_recent_posts( $num_posts ) {
+		global $wpdb;
+
+		$two_days_ago = date( 'Y-m-d', strtotime( '-2 days' ) );
+
+		/**
+		 * Filter post types to be included in news sitemap.
+		 *
+		 * @module sitemaps
+		 *
+		 * @since 3.9.0
+		 *
+		 * @param array $post_types Array with post types to include in news sitemap.
+		 */
+		$post_types = apply_filters(
+			'jetpack_sitemap_news_sitemap_post_types',
+			array( 'page', 'post' )
+		);
+
+		foreach ( (array) $post_types as $i => $post_type ) {
+			$post_types[ $i ] = $wpdb->prepare( '%s', $post_type );
+		}
+
+		$post_types_list = join( ',', $post_types );
+
+		return $wpdb->get_results(
```

Comment:
What are the pros/cons of querying the database directly v. using `get_posts()`.

If the news siteindex in generated on the fly for each request, it seems like there should be caching here. Thoughts?